Altering python app so that it behaves more like a regular python call

ivanimanishi opened the pull request:
In order to be able to easily make a python script capable of importing gaffer it is important variables like sys.argv have the same value regardless of whether they were called from  "python" or from "gaffer python".
We also want things like return values from the script to be preserved, and that the output from one or the other be similar.

My proposal here is to get rid of the traceback print out, as that was interfering with sys.exit calls, and replacing the return value with "1". That is the behaviour we get if we use python directly.

Also, I'm suggesting we "hack" the sys.argv variable value so that it matches what we'd get from a call to python, as explained in the code:
```python
		# trying to set the sys.argv to the same values as we would expect from a python call
		#
		# the expected gaffer python command line syntax is expected to be:
		#     gaffer python -file <script path> [-arguments <arguments>]
		#
		# given the limitations of the gaffer python app, we can assume that the equivalent python command syntax would be:
		#     python <script path> [<arguments>]
		#
		# python removes the "python" command from the start of the sys.argv list, making the first argument the script path
		# sys.argv should therefore be:
		#     [<script path>] + [<arguments>]
		#
		# sys.argv is assumed to be the base source of information for python argument parsing methods (argparse, getopt, ...)
```

This is changing the existing behaviour of the gaffer python app, but I'm not aware of any uses, so maybe it would be ok to not preserve full backward compatibility. But if that is an issue, we could create a different app with the changes I'm suggesting (possibly internal to IE).
Note that I'm still passing the argv variable (with an IECore.StringVectorData) as before, as that is not part of regular python.

There is another change that we may consider, which is to alter the `__name__` python variable.
When running python with a script, you get that set to `__main__`. Using the gaffer python app we get `__builtin__`. Maybe we should set it to `__main__`? But changing existing code to support both is not difficult, and there could be reasons to keep it as `__builtin__` that I'm not aware of.

ivanimanishi commented:
I like the idea of catching the SystemExit and returning that result as the result for the application. I'll look into that.
And, as the change in the sys.argv value is a requirement right now, I'll also look into implementing the threadsafe-argv-clobbering as well. If it turns out to be too complicated, we'd need to go with a IE-specific application.

johnhaddon commented:
Just to clarify, I'm OK with thread-unsafe argv clobbering in Gaffer for now if we stick in a comment about the potential issue. Since we don't currently use the python app inside the same processes as others, I think getting your other improvements into Gaffer is preferable over maintaining something internal to IE in this case.

ivanimanishi commented:
I've just updated the pull request with your suggestions.
About the thread-safe argv changes, I was able to get something to work, wrapping the list class using the __metaclass__ feature in python, but it seemed like something that could potentially be more problematic than the issue we were trying to address, specially given how unlikely it is that we'd have that app running on a thread.

One think to keep in mind is that even if the thread-safe sys.argv was implemented using a method similar to the one you described (using threading.local), it would potentially cause a generic python script to break if you actually wanted to change the sys.argv from a thread. Very, very unlikely, I know, but still a theoretical possibility...

Anyway, I've updated the comments to include a note that it is not thread-safe, and why we decided to ignore it.

johnhaddon commented:
Great - thanks Ivan. I didn't know that `finally` blocks also get executed from return statements inside the `try`, so I learned something. The metaclass thing sounds ingenious - good to know there's an option there should it become a real issue.

Just one small note for the future - it's very helpful to me if the first line of the commit message contains some context ("the python app" in this case). It really helps when compiling release notes and trawling back through the history to find problems.
